src/target/riscv/riscv-013_reg.c

```diff
@@ -85,33 +87,247 @@ static int riscv013_reg_set(struct reg *reg, uint8_t *buf)
+			/* exist */ true, riscv013_gdb_regno_reg_type(regno));
+}
+
+static int examine_xlen(struct target *target)
+{
+	RISCV_INFO(r);
+	unsigned int cmderr;
+
+	const uint32_t command = riscv013_access_register_command(target,
+			GDB_REGNO_S0, /* size */ 64, AC_ACCESS_REGISTER_TRANSFER);
+	int res = riscv013_execute_abstract_command(target, command, &cmderr);
+	if (res == ERROR_OK) {
+		r->xlen = 64;
+		return ERROR_OK;
+	}
+	if (res == ERROR_TIMEOUT_REACHED)
 		return ERROR_FAIL;
+	r->xlen = 32;
+
+	return ERROR_OK;
+}
+
+static int examine_vlenb(struct target *target)
+{
+	RISCV_INFO(r);
+
+	/* Reading "vlenb" requires "mstatus.vs" to be set, so "mstatus" should
+	 * be accessible.*/
+	int res = init_cache_entry(target, GDB_REGNO_MSTATUS);
+	if (res != ERROR_OK)
+		return res;
+
+	res = assume_reg_exist(target, GDB_REGNO_VLENB);
+	if (res != ERROR_OK)
+		return res;
+
+	riscv_reg_t vlenb_val;
+	if (riscv_reg_get(target, &vlenb_val, GDB_REGNO_VLENB) != ERROR_OK) {
+		if (riscv_supports_extension(target, 'V'))
+			LOG_TARGET_WARNING(target, "Couldn't read vlenb; vector register access won't work.");
+		r->vlenb = 0;
+		return riscv_reg_impl_set_exist(target, GDB_REGNO_VLENB, false);
+	}
+	/* As defined by RISC-V V extension specification:
+	 * https://github.com/riscv/riscv-v-spec/blob/2f68ef7256d6ec53e4d2bd7cb12862f406d64e34/v-spec.adoc?plain=1#L67-L72 */
+	const unsigned int vlen_max = 65536;
+	const unsigned int vlenb_max = vlen_max / 8;
+	if (vlenb_val > vlenb_max) {
+		LOG_TARGET_WARNING(target, "'vlenb == %" PRIu64
+				"' is greater than maximum allowed by specification (%u); vector register access won't work.",
+				vlenb_val, vlenb_max);
+		r->vlenb = 0;
+		return ERROR_OK;
+	}
+	assert(vlenb_max <= UINT_MAX);
+	r->vlenb = (unsigned int)vlenb_val;
+
+	LOG_TARGET_INFO(target, "Vector support with vlenb=%u", r->vlenb);
+	return ERROR_OK;
+}
+
+enum misa_mxl {
+	MISA_MXL_INVALID = 0,
+	MISA_MXL_32 = 1,
+	MISA_MXL_64 = 2,
+	MISA_MXL_128 = 3
+};
+
+enum misa_mxl xlen_to_mxl(unsigned int xlen)
+{
+	switch (xlen) {
+	case 32:
+		return MISA_MXL_32;
+	case 64:
+		return MISA_MXL_64;
+	case 128:
+		return MISA_MXL_128;
+	}
+	return MISA_MXL_INVALID;
+}
+
+static int check_misa_mxl(const struct target *target)
+{
+	RISCV_INFO(r);
+
+	if (r->misa == 0) {
+		LOG_TARGET_WARNING(target, "'misa' register is read as zero."
+				"OpenOCD will not be able to determine some hart's capabilities.");
+		return ERROR_OK;
+	}
+	const unsigned int xlen = riscv_xlen(target);
+	assert(xlen <= sizeof(riscv_reg_t) * CHAR_BIT);
+	assert(xlen >= 2);
+	riscv_reg_t misa_mxl_field = (riscv_reg_t)0x3 << (xlen - 2);
+	const unsigned int mxl = get_field(r->misa, misa_mxl_field);
+	/* RISC-V Debug Spec. requires "DXLEN" to be the widest. */
+	if (mxl == MISA_MXL_INVALID || xlen_to_mxl(xlen) < mxl) {
+		LOG_TARGET_ERROR(target,
+				"'misa.mxl' field's value of %u does not "
+				" coincide with 'xlen' of %u.",
+				mxl, xlen);
+		return ERROR_FAIL;
+	}
```

Comment:
1) If you like, the situation could be made more clear to the reader by producing different error messages (shown below).

2) Also it seems more natural to convert MXL to MXLEN and then make the comparison of the *XLEN values. That's because the privileged spec says `XLEN is never greater than MXLEN`.

3) While making changes here, it may be good to check for MXLEN = 128. If detected, report an error to the user, saying that OpenOCD currently does not support it. 

```suggestion
	/* RISC-V Debug Spec. requires "DXLEN" to be the widest. */
	if (mxl == MISA_MXL_INVALID) {
		LOG_TARGET_ERROR(target, "Invalid value of MXL: misa.mxl == 0");
		return ERROR_FAIL;
	}	
	if (xlen > mxl_to_xlen(mxl)) {
		LOG_TARGET_ERROR(target,
				"The detected XLEN (%u) exceeds the XLEN reported in misa.mxl field (%u)",
				...);
		return ERROR_FAIL;
	}
```

Reply:
1,2 - Addressed.

> it may be good to check for MXLEN = 128

AFAIU, this can not be done here: if `misa` is read assuming DXLEN is 64, when it is actually 128, `misa.mxl` will be read from `misa[61:63]`, which will be zero (or possibly non-zero in case of Priv. Spec 1.10).

Reply:
1,2 - Thank you.

3 - You're right. The detection of MXLEN=128 would have to be done differently than what I suggested (for instance by attempting a 128-bit abstract register read).